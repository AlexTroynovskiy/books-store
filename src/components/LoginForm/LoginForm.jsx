import React from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';

import { useFormik } from 'formik';
import * as Yup from 'yup';

import { TextField, Container } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import { ButtonComponent } from '../_shared/Button';

const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    height: '80vh',
    transform: 'translateY(-5%)',
  },
  form: {
    margin: '0 auto',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    width: '600px',
  },
  formField: {
    margin: '20px 0',
    width: '400px',
  },
  button: {
    marginTop: '20px',
  },
});

export const LoginForm = (props) => {
  const history = useHistory();

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
    },

    validationSchema: Yup.object().shape({
      email: Yup.string().email('Invalid email').required('Required'),
      password: Yup.string()
        .matches(/^(?=^.{6,}$)((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, 'Password invalid')
        .required('Required'),
    }),

    onSubmit: () => {
      props.changeStatus(true);
      history.push('/books');
    },
  });

  const classes = useStyles();

  return (
    <Container component="div" className={classes.root}>
      <form className={classes.form} onSubmit={formik.handleSubmit}>
        <TextField
          className={classes.formField}
          error={formik.errors.email && formik.touched.email}
          helperText={formik.touched.email && formik.errors.email}
          id="email"
          name="email"
          type="text"
          label="Email"
          variant="outlined"
          {...formik.getFieldProps('email')}
        />
        <TextField
          className={classes.formField}
          error={formik.errors.password && formik.touched.password}
          helperText={formik.touched.password && formik.errors.password}
          id="password"
          name="password"
          type="password"
          label="Password"
          variant="outlined"
          {...formik.getFieldProps('password')}
        />
        <ButtonComponent
          size="large"
          className={classes.button}
          type="submit"
          color="primary"
          variant="contained"
          disabled={!(formik.isValid && formik.dirty)}
        >
          Submit
        </ButtonComponent>
      </form>
    </Container>
  );
};

LoginForm.propTypes = {
  isLogged: PropTypes.bool.isRequired,
  changeStatus: PropTypes.func.isRequired,
};
