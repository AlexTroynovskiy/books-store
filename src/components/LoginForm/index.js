import { connect } from 'react-redux';

import { LoginForm } from './LoginForm';
import { changeStatus } from '../../redux/UserSpecific/actions';

const mapStateToProps = (state) => {
  return {
    isLogged: state.user.isLogged,
  };
};

const mapDispatchToProps = {
  changeStatus,
};

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
