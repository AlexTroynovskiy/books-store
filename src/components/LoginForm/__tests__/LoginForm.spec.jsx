import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { Router } from 'react-router';

import LoginFormConnect from '../';
import { LoginForm } from '../LoginForm';

describe('Component: LoginForm', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('Should be properly connected to the store', () => {
    const initialState = { isLogged: false, changeStatus: jest.fn() };
    const mockStore = configureStore();
    let store, wrapper;
    store = mockStore(initialState);

    beforeEach(() => {
      wrapper = shallow(
        <Provider store={store}>
          <LoginFormConnect {...initialState} />
        </Provider>
      );
    });

    test('To be in DOM', () => {
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Should have nested elements', () => {
    let wrapper;
    const props = { isLogged: false, changeStatus: jest.fn() };
    const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };

    test('To exist', () => {
      wrapper = shallow(
        <Router history={historyMock}>
          <LoginForm {...props} />
        </Router>
      );
      expect(wrapper.exists()).toBe(true);
    });

    test('To match snapshot', () => {
      wrapper = renderer
        .create(
          <Router history={historyMock}>
            <LoginForm {...props} />
          </Router>
        )
        .toJSON();

      expect(wrapper).toMatchSnapshot();
    });
  });
});
