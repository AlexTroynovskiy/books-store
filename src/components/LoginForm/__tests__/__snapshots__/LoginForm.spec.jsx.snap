// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: LoginForm Should have nested elements To match snapshot 1`] = `
<div
  className="form-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-root-1"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="email"
        id="email-label"
      >
        Email
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="email"
          name="email"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-60 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-62"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-root-1"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink={false}
        htmlFor="password"
        id="password-label"
      >
        Password
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input"
          disabled={false}
          id="password"
          name="password"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="password"
          value=""
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-60 MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-62"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained btn MuiButton-containedPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Submit
      </span>
    </button>
  </form>
</div>
`;
